enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -0,0 +1,326 @@
+the IP will get the internal certificate and wont validate against the external
+CA.
+
+If we were to include the node IP in the external certificate we also get an
+immediate match in the mapping, and the hostname verification also works ok.
+All internal connections (those coming from pods through the service network
+IP) will be broken when trying to validate hostname, as the service IP is
+not included. Even if it was included, validation with the internal CA would
+still fail.
+
+Including the IP in the current situation means choosing what to lose:
+internal or external connectivity. Since internal connectivity is mandatory,
+the only option is to use names for external connectivity. MicroShift will run
+in environments where hostname (or DNS) access may not be guaranteed. Customers
+have given us the requirement of accessing the cluster via IP addresses.
+
+### Merging certificates
+Since having the node IP coexist with trust domain segregation requires not
+including it for external access, there are not too many options left for
+fulfilling the requirements.
+
+Like K3s, which is a competing project, the approach would be to merge all API
+server certificates into a single one, including service names, service IP,
+node IP, node names, and additional SANs from the configuration. The
+certificate matching problem vanishes, as well as the hostname validations and
+different CAs, as it would be a single one.
+
+This approach removes trust domains when accessing API server, like OCP does,
+but the use cases require IP connectivity and authentication from anywhere.
+
+
+### Workflow Description
+Workflow is transparent for the user, as certificates generate themselves upon
+restarting MicroShift.
+
+#### Variation [optional]
+N/A
+
+### API Extensions
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+N/A
+
+### Risks and Mitigations
+Removal of trust domains and different approach than the one in OpenShift.
+
+Having a single certificate to connect to API server means anyone can
+connect to it, and not having oauth requires extra care when configuring
+cluster roles. This, however, is not too different than the current approach.
+
+Not implementing this change makes MicroShift unable to meet requirements.
+
+### Drawbacks
+See risks above.
+
+## Design Details
+N/A
+
+### Open Questions [optional]
+N/A
+
+### Test Plan
+New e2e tests should be introduced to change:
+* Hostname.
+* IP.
+* SAN.
+
+And verify connectivity and authentication is still working from any client,
+external or internal.
+
+### Graduation Criteria
+N/A
+
+#### Dev Preview -> Tech Preview
+N/A
+
+#### Tech Preview -> GA
+N/A
+
+#### Removing a deprecated feature
+N/A
+
+### Upgrade / Downgrade Strategy
+N/A
+
+### Version Skew Strategy
+N/A
+
+### Operational Aspects of API Extensions
+N/A
+
+#### Failure Modes
+N/A
+
+#### Support Procedures
+N/A
+
+## Implementation History
+N/A
+
+## Alternatives
+* Have several IPs in the node.
```

Comment:
Let's separate these things, because they're related but not tightly.

We have a requirement that we cannot ask the user to give us 2 IPs on their network to run 1 host with MicroShift. We could use an internal network if we manage the IPs in a way that do not conflict with other networks in the environment.

We have a requirement that we run the conformance tests, including a test that checks for 2 nodes. That means that whatever we come up with _does_ need to work if there is a second node. We do not expect to support that configuration in production settings, though.